subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipCachedTaskExecuter.java

```diff
@@ -47,81 +48,69 @@
     private final TaskExecuter delegate;
     private final TaskOutputsGenerationListener taskOutputsGenerationListener;
     private final TaskOutputOriginFactory taskOutputOriginFactory;
+    private final TaskCachingReasonsListener taskCachingReasonsListener;
     private BuildCache cache;
 
-    public SkipCachedTaskExecuter(TaskOutputOriginFactory taskOutputOriginFactory, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskExecuter delegate) {
+    public SkipCachedTaskExecuter(TaskOutputOriginFactory taskOutputOriginFactory, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskCachingReasonsListener taskCachingReasonsListener, TaskExecuter delegate) {
         this.taskOutputOriginFactory = taskOutputOriginFactory;
         this.buildCacheConfiguration = buildCacheConfiguration;
         this.packer = packer;
         this.taskOutputsGenerationListener = taskOutputsGenerationListener;
         this.delegate = delegate;
+        this.taskCachingReasonsListener = taskCachingReasonsListener;
     }
 
     @Override
     public void execute(final TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
         final Timer clock = Timers.startTimer();
 
         final TaskOutputsInternal taskOutputs = task.getOutputs();
-
-        boolean cacheEnabled;
+        boolean cacheable;
         try {
-            cacheEnabled = taskOutputs.isCacheEnabled();
+            TaskCaching taskCaching = taskOutputs.getTaskCaching();
+            cacheable = taskCaching.isCacheable();
+            state.setCacheable(cacheable);
+            taskCachingReasonsListener.taskCacheable(task, taskCaching);
```

Comment:
That would mean that we could drop the listener altogether and have a client use a `TaskExecutionListener`. By listening to `beforeExecute(Task)` the client would be able to obtain `task.outputs.caching`. I feels to me that we probably want to do that given that I cannot come up with a good name for the method of the listener. Only something like `taskCachingEvaluated` would come to mind. We also could add the whole immutable `TaskOutputCaching` object to `TaskStateInternal` removing `TaskStateInternal.isCacheable()` in the process.
What is your take on this?